safe_transaction_service/history/models.py

```diff
@@ -297,58 +310,145 @@ def update_with_block_and_receipt(
+        topic = HexBytes(event_data["topics"][0])
+        expected_topic = HexBytes(ERC20_721_TRANSFER_TOPIC)
+        if topic != expected_topic:
+            raise ValueError(
+                f"Not supported EventData, topic {topic.hex()} does not match expected {expected_topic.hex()}"
             )
-        return queryset
 
-    def erc20_events(
-        self, token_address: Optional[str] = None, address: Optional[str] = None
-    ):
-        return self.erc20_and_721_events(
-            token_address=token_address, address=address
-        ).filter(arguments__has_key="value")
+        return {
+            "ethereum_tx_id": event_data["transactionHash"],
+            "log_index": event_data["logIndex"],
+            "address": event_data["address"],
+            "_from": event_data["args"]["from"],
+            "to": event_data["args"]["to"],
+        }
+
+    @classmethod
+    def from_decoded_event(cls, event_data: EventData):
+        raise NotImplementedError
+
+    @property
+    def created(self):
+        return self.ethereum_tx.block.timestamp
+
+
+class ERC20TransferQuerySet(TokenTransferQuerySet):
+    def token_txs(self):
+        return self.annotate(
+            _value=F("value"),
+            transaction_hash=F("ethereum_tx_id"),
+            block_number=F("ethereum_tx__block_id"),
+            execution_date=F("ethereum_tx__block__timestamp"),
+            _token_id=RawSQL("NULL::numeric", ()),
+            token_address=F("address"),
+        ).order_by("-ethereum_tx__block_id")
 
-    def erc721_events(
-        self, token_address: Optional[str] = None, address: Optional[str] = None
-    ):
-        # TODO Sql with tokens registered as NFT
-        return self.erc20_and_721_events(
-            token_address=token_address, address=address
-        ).filter(arguments__has_key="tokenId")
 
+class ERC20Transfer(TokenTransfer):
+    objects = TokenTransferManager.from_queryset(ERC20TransferQuerySet)()
+    value = Uint256Field()
+
+    class Meta:
+        verbose_name = "ERC20 Transfer"
+        verbose_name_plural = "ERC20 Transfers"
+
+    def __str__(self):
+        return f"Token Transfer from={self._from} to={self.to} value={self.value}"
+
+    @classmethod
+    def from_decoded_event(cls, event_data: EventData) -> Union["ERC20Transfer"]:
+        """
+        Does not create the model, as it requires that `ethereum_tx` exists
+
+        :param event_data:
+        :return: `ERC20Transfer`
+        :raises: ValueError
+        """
+
+        parameters = cls._prepare_parameters_from_decoded_event(event_data)
+
+        if "value" in event_data["args"]:
+            parameters["value"] = event_data["args"]["value"]
+            return ERC20Transfer(**parameters)
+        else:
+            raise ValueError(
+                f"Not supported EventData, `value` not present {event_data}"
+            )
+
+
+class ERC721TransferManager(TokenTransferManager):
+    # TODO Optimize this
```

Comment:
Is this `TODO` still valid? 😅

Reply:
Yes, it is. I would like to find a way to do this directly on SQL